### PR TITLE
Feat/remove shiftfs

### PR DESCRIPTION
**Pull request by gabrielmougard**
Removing ShiftFS

### CODE REVIEW

**Comment by simondeziel** [comment]
Should https://documentation.ubuntu.com/lxd/en/latest/reference/instance_options/#instance-security:security.syscalls.intercept.mount.shift be dropped (doc and config key)?

**Comment by gabrielmougard** [comment]
> Should https://documentation.ubuntu.com/lxd/en/latest/reference/instance_options/#instance-security:security.syscalls.intercept.mount.shift be dropped (doc and config key)?
> 
> Searching our doc also turn up other cases of `shiftfs` that probably need to be updated/replaced.

I don't think so, we still are using it to check whether a mount syscall needs shifting:

```go
func (s *Server) MountSyscallShift(c Instance, path string) idmap.IdmapStorageType {
	if shared.IsTrue(c.ExpandedConfig()["security.syscalls.intercept.mount.shift"]) {
		diskIdmap, err := c.DiskIdmap()
		if err != nil {
			return idmap.IdmapStorageNone
		}

		if diskIdmap == nil {
			return c.IdmappedStorage(path, "none")
		}
	}

	return idmap.IdmapStorageNone
}
```

**Comment by simondeziel** [comment]
> > Should https://documentation.ubuntu.com/lxd/en/latest/reference/instance_options/#instance-security:security.syscalls.intercept.mount.shift be dropped (doc and config key)?
> > Searching our doc also turn up other cases of `shiftfs` that probably need to be updated/replaced.
> 
> I don't think so, we still are using it to check whether a mount syscall needs shifting:

The help text literally mentions `shiftfs`.

**Comment by gabrielmougard** [comment]
@simondeziel updated

**Comment by simondeziel** [comment]
https://canonical-ubuntu-documentation-library--12516.com.readthedocs.build/lxd/en/12516/search/?q=shiftfs&check_keywords=yes&area=default shows 2 hits. For the API page, I think replacing or adding the `idmapped mounts` would make sense for all 3 occurrences of `shiftfs`. **If** `shiftfs` is replaced by only `idmapped mounts`, we should probably drop it from the FAQ page too.

**Review by simondeziel — DISMISSED** [review]
LGTM, thanks!

**Review by tomponline — CHANGES_REQUESTED** [review]
Im not sure we should be removing test_container_devices_disk_shift

File: test/suites/container_devices_disk.sh [hunk 1]
```diff
@@ -14,65 +13,6 @@ test_container_devices_disk() {
   lxc delete -f foo
 }
 
-test_container_devices_disk_shift() {
```
[the remark this answers is not in the document]
Comment (mihalicyn) [reply test/suites/container_devices_disk.sh:17]: yes, I think that we need to keep this test and modify a condition `  if ! grep -q shiftfs /proc/filesystems || [ -n "${LXD_SHIFTFS_DISABLE:-}" ]; then` to be something like `  if [ -n "${LXD_IDMAPPED_MOUNTS_DISABLE:-}" ]; then`

**Comment by gabrielmougard** [comment]
@mihalicyn Can you confirm for the test please?

**Comment by tomponline** [comment]
@gabrielmougard yes a kernel and distribution with ZFS >= 2.2

**Comment by tomponline** [comment]
@mihalicyn has an idea on how we can detect if the pool supports idmapped mounts he will let you know.

**Review by tomponline — APPROVED** [review]

**Pull request merged**